### PR TITLE
Projects Syncer

### PR DESCRIPTION
**Pull request by gazarenkov**
WorkspaceProjectsSyncer is an abstract class to be implemented by concrete entities which hold Projects configuration. WorkspaceHolder is an implementation which uses Workspace database remotely, on Ws Master. 

@vparfonov please check it out

### CODE REVIEW

File: plugins/plugin-maven/che-plugin-maven-server/src/test/java/org/eclipse/che/api/project/server/TestWorkspaceHolder.java [hunk 1]
```diff
@@ -10,46 +10,38 @@
...
-    }
-
-    @Override
-    void removeProjects(Collection<RegisteredProject> projects) throws ServerException {
-        projects.stream()
-                .filter(project -> !project.isDetected())
-                .forEach(workspace::removeProject);
-    }
-}
+//public class TestWorkspaceHolder extends WorkspaceHolder {
```
Comment (benoitf) [inline plugins/plugin-maven/che-plugin-maven-server/src/test/java/org/eclipse/che/api/project/server/TestWorkspaceHolder.java]: why the test class is commented ? test should work. If it's dead code, then it needs to be removed

**Comment by vparfonov** [comment]
@skabashnyuk

File: plugins/plugin-maven/che-plugin-maven-server/src/test/java/org/eclipse/che/plugin/maven/server/BaseTest.java [hunk 2]
```diff
@@ -228,4 +232,51 @@ protected TestProjectType() {
         }
     }
 
+//    protected static class MavenProjectType extends ProjectTypeDef {
+//
+//        protected MavenProjectType() {
+//            super("maven", "maven", true, true);
```
Comment (skabashnyuk) [inline plugins/plugin-maven/che-plugin-maven-server/src/test/java/org/eclipse/che/plugin/maven/server/BaseTest.java]: cleanup

File: plugins/plugin-maven/che-plugin-maven-server/src/test/java/org/eclipse/che/plugin/maven/server/projecttype/handler/MavenProjectResolverTest.java [hunk 3]
```diff
@@ -165,18 +163,60 @@ private void createModule(Path parent, String moduleName) throws Exception {
...
+        }
+
+
     }
+
+
+
+//
+//    protected static class TestWorkspaceHolder extends WorkspaceHolder {
+//
```
Comment (skabashnyuk) [inline plugins/plugin-maven/che-plugin-maven-server/src/test/java/org/eclipse/che/plugin/maven/server/projecttype/handler/MavenProjectResolverTest.java]: cleanup

File: plugins/plugin-maven/che-plugin-maven-server/src/test/java/org/eclipse/che/plugin/maven/server/projecttype/handler/MavenProjectResolverTest.java [hunk 4]
```diff
@@ -165,18 +163,60 @@ private void createModule(Path parent, String moduleName) throws Exception {
         Files.write(pom, pomJar.getBytes());
     }
 
-    protected static class TestWorkspaceHolder extends WorkspaceHolder {
+    protected static class TestWorkspaceHolder extends WorkspaceProjectsSyncer {
```
Comment (skabashnyuk) [inline plugins/plugin-maven/che-plugin-maven-server/src/test/java/org/eclipse/che/plugin/maven/server/projecttype/handler/MavenProjectResolverTest.java:170]: What is it? javadoc? shall we extract it as separate class?

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java [hunk 5]
```diff
@@ -99,6 +100,8 @@ public ProjectManager(VirtualFileSystemProvider vfsProvider,
         this.importers = importers;
         this.fileWatchNotifier = fileWatcherNotificationHandler;
         this.fileWatcher = fileTreeWatcher;
+        this.workspaceProjectsHolder = workspaceProjectsHolder;
+
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java:104]: do we really need this  extra space?

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java [hunk 6]
```diff
@@ -246,6 +249,8 @@ public RegisteredProject createProject(ProjectConfig projectConfig, Map<String,
             throw e;
         }
 
+        workspaceProjectsHolder.sync(projectRegistry);
+
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java:253]: do we really need this  extra space?

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java [hunk 7]
```diff
@@ -287,6 +292,8 @@ public RegisteredProject updateProject(ProjectConfig newConfig) throws Forbidden
 
         final RegisteredProject project = projectRegistry.putProject(newConfig, baseFolder, true, false);
 
+        workspaceProjectsHolder.sync(projectRegistry);
+
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java:296]: do we really need this  extra space?

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectRegistry.java [hunk 8]
```diff
@@ -75,9 +75,9 @@ public ProjectRegistry(WorkspaceHolder workspaceHolder,
 
     @PostConstruct
     public void initProjects() throws ConflictException, NotFoundException, ServerException, ForbiddenException {
-        final Workspace workspace = workspaceHolder.getWorkspace();
+        //final Workspace workspace = workspaceHolder.getWorkspace();
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectRegistry.java:78]: cleanup

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/WorkspaceHolder.java [hunk 9]
```diff
@@ -45,21 +32,24 @@
...
     @Inject
     public WorkspaceHolder(@Named("api.endpoint") String apiEndpoint,
                            @Named("env.CHE_WORKSPACE_ID") String workspaceId,
                            HttpJsonRequestFactory httpJsonRequestFactory) throws ServerException {
+
         this.apiEndpoint = apiEndpoint;
         this.httpJsonRequestFactory = httpJsonRequestFactory;
 
+        this.workspaceId = workspaceId;
+
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/WorkspaceHolder.java:52]: do we really need this  extra space?

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/WorkspaceHolder.java [hunk 10]
```diff
@@ -68,48 +58,42 @@ public WorkspaceHolder(@Named("api.endpoint") String apiEndpoint,
...
 
-        this.workspace = new UsersWorkspaceImpl(workspaceDto(workspaceId));
     }
 
-    @VisibleForTesting
-    protected WorkspaceHolder(WorkspaceDto workspace) throws ServerException {
-        this.workspace = new UsersWorkspaceImpl(workspace);
+
+
+
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/WorkspaceHolder.java:65]: cleanup extra space

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/WorkspaceHolder.java [hunk 11]
```diff
@@ -68,48 +58,42 @@ public WorkspaceHolder(@Named("api.endpoint") String apiEndpoint,
...
-     * @return workspace object
-     */
-    public Workspace getWorkspace() {
-        return this.workspace;
+    @Override
+    public String getWorkspaceId() {
+        return workspaceId;
     }
 
+
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/WorkspaceHolder.java:77]: do we really need this  extra space?

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/WorkspaceProjectsSyncer.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+        // check on removed
+        List <ProjectConfig> removed = new ArrayList<>();
+        for(ProjectConfig r  : remote) {
+            if(projectRegistry.getProject(r.getPath()) == null)
+                removed.add(r);
+        }
+
+        for(ProjectConfig r : removed)
+            removeProject(r);
+
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/WorkspaceProjectsSyncer.java:44]: cleanup extra space

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/WorkspaceProjectsSyncer.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+                        updateProject(config);
+                        found = true;
+                    }
+                }
+
+                if(!found)
+                    addProject(config);
+
+                project.setSync();
+
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/WorkspaceProjectsSyncer.java:71]: cleanup extra space

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/WorkspaceProjectsSyncer.java [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+            }
+
+
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/WorkspaceProjectsSyncer.java:74]: cleanup extra space

File: wsagent/che-core-api-project/src/test/java/org/eclipse/che/api/project/server/WsAgentTestBase.java [hunk 15]
```diff
@@ -118,50 +115,53 @@ public void setUp() throws Exception {
...
-            super(DtoFactory.newDto(WorkspaceDto.class)
-                            .withId("id")
-                            .withConfig(DtoFactory.newDto(WorkspaceConfigDto.class)
-                                                  .withName("name")
-                                                  .withProjects(projects)));
+        protected TestWorkspaceHolder(List<ProjectConfig> projects) throws ServerException {
+            for (ProjectConfig p : projects) {
+                this.projects.put(p.getPath(), p);
+            }
+
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/test/java/org/eclipse/che/api/project/server/WsAgentTestBase.java:139]: cleanup extra space

File: wsagent/che-core-api-project/src/test/java/org/eclipse/che/api/project/server/WsAgentTestBase.java [hunk 16]
```diff
@@ -118,50 +115,53 @@ public void setUp() throws Exception {
...
+                                importerRegistry, fileWatcherNotificationHandler, fileTreeWatcher, wsHolder);
         pm.initWatcher();
     }
 
 
-    protected static class TestWorkspaceHolder extends WorkspaceHolder {
+    protected static class TestWorkspaceHolder extends WorkspaceProjectsSyncer {
 
-        //ArrayList <RegisteredProject> updatedProjects = new ArrayList<>();
+        private HashMap<String, ProjectConfig> projects = new HashMap<>();
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/test/java/org/eclipse/che/api/project/server/WsAgentTestBase.java:129]: private Map<String, ProjectConfig> projects = new HashMap<>();
?

**Pull request merged**